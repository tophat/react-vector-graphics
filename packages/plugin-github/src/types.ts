import type { Octokit } from "@octokit/rest"

export type PluginGitHubOptions = {
    'github/api': Octokit

    'github/base': string
    'github/commitCreate'?: string
    'github/commitDelete'?: string
    'github/commitUpdate'?: string
    'github/folderPath': string
    'github/head': string
    'github/owner': string
    'github/repo': string
    'assets/fileExt': string
    'assets/globPattern': string
    'assets/nameScheme': string
    'assets/outputPath': string
}
