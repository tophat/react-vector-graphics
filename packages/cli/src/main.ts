import * as path from 'path'

import { getLogger, default as rvgCore } from '@react-vector-graphics/core'
import yargs from 'yargs'

import * as conf from './config'

import type { Config, Logger } from '@react-vector-graphics/types'
import type { CosmiconfigResult } from 'cosmiconfig/dist/types'

const configResultHandler =
    (config: Config, logger: Logger) =>
    (result: CosmiconfigResult): void => {
        if (result?.filepath && !result.isEmpty) {
            logger.info(`Loaded config from ${result.filepath}`)
            Object.assign(config, result.config)
            const configDir = path.dirname(result.filepath)
            if (configDir !== process.cwd()) {
                logger.info(`Changing current directory to ${configDir}`)
                process.chdir(configDir)
            }
        } else if (result?.isEmpty) {
            logger.info(`Skipping empty config at ${result.filepath}`)
        } else {
            logger.info('No config loaded')
        }
    }

const getConfig = async (config: Config, logger: Logger): Promise<Config> => {
    const currentDir = process.cwd()
    const handleConfig = configResultHandler(config, logger)
    logger.info(`Searching for config in ${currentDir}`)
    const result = await conf.find(currentDir)
    handleConfig(result)

    logger.info('Start getting cli options')
    const { argv } = yargs(process.argv)
        .usage('Usage: $0 --config [config]')
        .option('config', {
            default: result?.filepath,
            describe: 'Config file path',
            type: 'string',
        })
        .demandOption(['config'])

    if (result?.filepath !== argv.config) {
        await conf.load(argv.config).then(handleConfig)
    }
    return config
}

export const run = async (): Promise<void> => {
    const defaultConfig = conf.init()
    const logger = getLogger()
    const config = await getConfig(defaultConfig, logger)
    if (!config.plugins?.length) {
        logger.info('Setting the minimal plugin configuration')
        const assetPlugin = '@react-vector-graphics/plugin-assets'
        config.plugins = [assetPlugin, '@svgr/plugin-jsx', assetPlugin]
    }
    await rvgCore({ config, logger })
}
