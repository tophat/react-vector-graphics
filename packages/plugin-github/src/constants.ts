export const OPTIONS = {
    API: `github/api`,
    BASE: `github/base`,
    COMMIT_CREATE_PATTERN: `github/commitCreate`,
    COMMIT_DELETE_PATTERN: `github/commitDelete`,
    COMMIT_UPDATE_PATTERN: `github/commitUpdate`,
    FOLDER_PATH: `github/folderPath`,
    HEAD: `github/head`,
    OWNER: `github/owner`,
    REPO: `github/repo`,
    FILE_EXT: 'assets/fileExt', // @react-vector-graphics/plugin-assets
    GLOB_PATTERN: 'assets/globPattern', // @react-vector-graphics/plugin-assets
    NAME_SCHEME: 'assets/nameScheme', // @react-vector-graphics/plugin-assets
    OUTPUT_PATH: 'assets/outputPath', // @react-vector-graphics/plugin-assets
} as const

export const STATE = {
    COMPONENT_FILES: 'assets/componentFiles', // @react-vector-graphics/plugin-assets
    COMPONENT_NAME: 'componentName', // @svgr/plugin-jsx
    COMPONENT_NAME_OLD: `github/componentNameOld`,
    DIFF_TYPE: `github/diffType`,
    FILE_PATH: 'filePath', // @svgr/plugin-jsx
} as const

export const STATUSES = {
    ADDED: 'added',
    MODIFIED: 'modified',
    REMOVED: 'removed',
    RENAMED: 'renamed',
} as const

export const COMMIT_MESSAGE_PLACEHOLDER = '${message}'

export const COMMIT_MESSAGE_PATTERNS = {
    CREATE: `feat: ${COMMIT_MESSAGE_PLACEHOLDER}`,
    DELETE: `refactor: ${COMMIT_MESSAGE_PLACEHOLDER}\n\nBREAKING CHANGE: ${COMMIT_MESSAGE_PLACEHOLDER}`,
    UPDATE: `fix: ${COMMIT_MESSAGE_PLACEHOLDER}`,
}

export const EMPTY_SVG = '<svg/>'
