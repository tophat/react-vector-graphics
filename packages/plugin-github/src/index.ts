export * from './constants'
export { run as default } from './main'
export type { PluginGitHubOptions } from './types'
