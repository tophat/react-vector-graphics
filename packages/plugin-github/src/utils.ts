import { type Logger } from '@react-vector-graphics/types'

// https://github.com/octokit/rest.js/issues/1971
type WithContent<T> = T & { content: Buffer | string }
export const withContent = <T>(data: T): WithContent<T> =>
    data as WithContent<T>

// https://github.com/isaacs/minimatch/issues/30
export const normaliseGlob = (globPattern: string): string => {
    if (globPattern.startsWith('./')) {
        return globPattern.substring(2)
    }
    return globPattern
}

export const replaceAll = (
    str: string,
    search: string,
    replacement: string,
): string => str.split(search).join(replacement)

export const fromBase64 = (data: string): string =>
    Buffer.from(data, 'base64').toString()

export const toBase64 = (data: string): string =>
    Buffer.from(data).toString('base64')

export const eagerPromises = (
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    promises: Promise<any>[],
    logger: Logger = console,
): // eslint-disable-next-line @typescript-eslint/no-explicit-any
Promise<any> => Promise.all(promises.map((p) => p.catch(logger.error)))
