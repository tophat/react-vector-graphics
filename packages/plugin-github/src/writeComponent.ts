import * as path from 'path'

import { type Octokit } from '@octokit/rest'
import { type Logger } from '@react-vector-graphics/types'

import {
    COMMIT_MESSAGE_PATTERNS,
    COMMIT_MESSAGE_PLACEHOLDER,
    OPTIONS,
    STATE,
    STATUSES,
} from './constants'
import {
    eagerPromises,
    fromBase64,
    replaceAll,
    toBase64,
    withContent,
} from './utils'

const ensureArray = <T>(o: T | T[]): T[] => (Array.isArray(o) ? o : [o])

const removeIconFiles = async (
    githubApi: Octokit,
    githubParams: { head: string; owner: string; repo: string },
    componentName: string,
    componentPath: string,
    commitMessagePattern: string = COMMIT_MESSAGE_PATTERNS.DELETE,
): Promise<void> => {
    const deleteMessage = `remove ${componentName}`
    const message = replaceAll(
        commitMessagePattern,
        COMMIT_MESSAGE_PLACEHOLDER,
        deleteMessage,
    )
    const { data: results } = await githubApi.repos.getContent({
        owner: githubParams.owner,
        path: componentPath,
        ref: githubParams.head,
        repo: githubParams.repo,
    })
    await eagerPromises(
        ensureArray(results).map(
            ({ path, sha }: { path: string; sha: string }) =>
                githubApi.repos.deleteFile({
                    branch: githubParams.head,
                    message,
                    owner: githubParams.owner,
                    path,
                    repo: githubParams.repo,
                    sha,
                }),
        ),
    )
}

const addOrModifyIconFile = async (
    githubApi: Octokit,
    githubParams: { head: string; owner: string; repo: string },
    componentName: string,
    fileName: string,
    filePath: string,
    fileContents: string,
    commitMessagePatternCreate: string = COMMIT_MESSAGE_PATTERNS.CREATE,
    commitMessagePatternUpdate: string = COMMIT_MESSAGE_PATTERNS.UPDATE,
    logger: Logger = console,
): Promise<void> => {
    let fileSha
    let fileContentsOld
    try {
        const { data } = await githubApi.repos.getContent({
            owner: githubParams.owner,
            path: filePath,
            ref: githubParams.head,
            repo: githubParams.repo,
        })
        if (Array.isArray(data)) {
            return void logger.info('Path is folder, skipping', filePath)
        }
        fileSha = data.sha
        fileContentsOld = fromBase64(withContent(data).content.toString())
    } catch (e) {
        logger.error(`${e}: ${filePath}`)
    }
    if (fileContentsOld === fileContents) {
        return void logger.info('No changes, skipping file', filePath)
    }
    const message = fileSha
        ? replaceAll(
              commitMessagePatternUpdate,
              COMMIT_MESSAGE_PLACEHOLDER,
              `modify ${componentName} ${path.basename(fileName)}`,
          )
        : replaceAll(
              commitMessagePatternCreate,
              COMMIT_MESSAGE_PLACEHOLDER,
              `add ${componentName} ${path.basename(fileName)}`,
          )
    await githubApi.repos.createOrUpdateFileContents({
        branch: githubParams.head,
        content: toBase64(fileContents),
        message,
        owner: githubParams.owner,
        path: filePath,
        repo: githubParams.repo,
        sha: fileSha,
    })
}

const writeComponent = async ({
    github: { api: githubApi, ...githubParams },
    ...params
}: {
    assetFile: string
    code: string
    commitMessagePatterns?: {
        create?: string
        delete?: string
        update?: string
    }
    componentName?: string
    componentNameOld?: string
    componentFiles: { [fileName: string]: string }
    diffType: string
    fileExt?: string
    folderPath: string
    github: {
        api: Octokit
        base: string
        head: string
        owner: string
        repo: string
    }
    logger?: Logger
    outputPath?: string
}): Promise<void> => {
    if (!params.componentName) {
        return params.logger?.warn(
            `No '${STATE.COMPONENT_NAME}' provided for '${params.assetFile}'.`,
        )
    }
    if (!params.outputPath) {
        return params.logger?.warn(`No '${OPTIONS.OUTPUT_PATH}' provided.`)
    }
    if (!params.fileExt) {
        params.logger?.warn(`No '${OPTIONS.FILE_EXT}' provided.`)
    }
    // gather files
    const componentFiles = Object.entries(params.componentFiles)
    const singleFile = componentFiles.length === 0
    const pathToFolder = path.join(
        params.folderPath,
        params.outputPath,
        singleFile ? '' : params.componentName,
    )
    const componentFileName =
        (singleFile ? params.componentName : 'index') +
        (params.fileExt ? `.${params.fileExt}` : '')
    componentFiles.push([componentFileName, params.code])
    const componentFilePath = path.join(pathToFolder, componentFileName)
    // commit file changes
    const pendingPromises = []
    if (params.diffType === STATUSES.REMOVED) {
        pendingPromises.push(
            removeIconFiles(
                githubApi,
                githubParams,
                params.componentName,
                singleFile ? componentFilePath : pathToFolder,
                params.commitMessagePatterns?.delete,
            ),
        )
    } else {
        // added, modified or renamed
        for (const [fileName, fileContents] of componentFiles) {
            const filePath = path.normalize(path.join(pathToFolder, fileName))
            pendingPromises.push(
                addOrModifyIconFile(
                    githubApi,
                    githubParams,
                    params.componentName,
                    fileName,
                    filePath,
                    fileContents,
                    params.commitMessagePatterns?.create,
                    params.commitMessagePatterns?.update,
                    params.logger,
                ),
            )
        }
        if (params.diffType === STATUSES.RENAMED && params.componentNameOld) {
            const oldPathToFolder = path.join(
                params.folderPath,
                params.outputPath,
                singleFile ? '' : params.componentNameOld,
            )
            const oldPathToFile = path.join(
                oldPathToFolder,
                singleFile ? params.componentNameOld : 'index',
            )
            const oldComponentFilePath = params.fileExt
                ? `${oldPathToFile}.${params.fileExt}`
                : oldPathToFile
            pendingPromises.push(
                removeIconFiles(
                    githubApi,
                    githubParams,
                    params.componentNameOld,
                    singleFile ? oldComponentFilePath : oldPathToFolder,
                    params.commitMessagePatterns?.delete,
                ),
            )
        }
    }
    await eagerPromises(pendingPromises)
}

export default writeComponent
