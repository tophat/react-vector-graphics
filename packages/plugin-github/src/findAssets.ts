import * as path from 'path'

import { pathToName } from '@react-vector-graphics/utils'
import minimatch from 'minimatch'

import { EMPTY_SVG, STATE, STATUSES } from './constants'
import { fromBase64, normaliseGlob, toBase64, withContent } from './utils'

import type { Octokit } from '@octokit/rest'
import type {
    NamingScheme,
    PluginParams,
    State,
} from '@react-vector-graphics/types'

const findAssets = async ({
    folderPath = '',
    github: { api: githubApi, base = 'master', ...githubParams },
    ...params
}: {
    folderPath: string
    github: {
        api: Octokit
        base: string
        head: string
        owner: string
        repo: string
    }
    globPattern: string
    nameScheme: NamingScheme
    state: Omit<State, 'componentName'> & { componentName?: string }
}): Promise<PluginParams[]> => {
    const compareCommitsResult = await githubApi.repos.compareCommits({
        ...githubParams,
        base,
    })
    const svgFiles =
        compareCommitsResult.data.files?.filter((file) => {
            const isInFolder = file.filename.startsWith(folderPath)
            if (!isInFolder) return false
            const relPath = path.relative(folderPath, file.filename)
            return minimatch(relPath, normaliseGlob(params.globPattern))
        }) ?? []
    const pluginParams = svgFiles.map(async (file): Promise<PluginParams> => {
        const filePath = path.relative(folderPath, file.filename)
        const { data } =
            file.status === STATUSES.REMOVED
                ? { data: { content: toBase64(EMPTY_SVG) } }
                : await githubApi.repos.getContent({
                      ...githubParams,
                      base,
                      path: file.filename,
                      ref: githubParams.head,
                  })
        const { content } = withContent(Array.isArray(data) ? {} : data)
        if (!content) {
            throw new Error(`Could not get contents for ${file.filename}`)
        }
        return {
            code: fromBase64(content.toString()),
            state: Object.assign(
                {
                    [STATE.COMPONENT_NAME]: pathToName(
                        filePath,
                        params.nameScheme,
                    ),
                    [STATE.COMPONENT_NAME_OLD]:
                        file.previous_filename &&
                        pathToName(
                            path.relative(folderPath, file.previous_filename),
                            params.nameScheme,
                        ),
                    [STATE.DIFF_TYPE]: file.status,
                    [STATE.FILE_PATH]: filePath,
                },
                params.state,
            ) as State,
        }
    })
    return Promise.all(pluginParams)
}

export default findAssets
