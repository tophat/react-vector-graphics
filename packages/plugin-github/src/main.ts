import { OPTIONS, STATE } from './constants'
import findAssets from './findAssets'
import writeComponent from './writeComponent'
import type { PluginGitHubOptions } from './types'
import type { NamingScheme, Plugin, State } from '@react-vector-graphics/types'

export const run: Plugin = async (code, config, state, logger) => {
    const options: Partial<PluginGitHubOptions> = config.options ?? {}

    const github = {
        api: options?.[OPTIONS.API],
        base: options?.[OPTIONS.BASE],
        head: options?.[OPTIONS.HEAD],
        owner: options?.[OPTIONS.OWNER],
        repo: options?.[OPTIONS.REPO],
    }
    const sharedArgs = {
        folderPath: options?.[OPTIONS.FOLDER_PATH],
        github,
    }
    if (code) {
        const opts = {
            ...sharedArgs,
            assetFile: state[STATE.FILE_PATH as keyof State],
            code,
            commitMessagePatterns: {
                create: options?.[OPTIONS.COMMIT_CREATE_PATTERN],
                delete: options?.[OPTIONS.COMMIT_DELETE_PATTERN],
                update: options?.[OPTIONS.COMMIT_UPDATE_PATTERN],
            },
            componentFiles: state[STATE.COMPONENT_FILES as keyof State] ?? {},
            componentName: state[STATE.COMPONENT_NAME as keyof State],
            componentNameOld: state[STATE.COMPONENT_NAME_OLD as keyof State],
            diffType: state[STATE.DIFF_TYPE as keyof State],
            fileExt: options?.[OPTIONS.FILE_EXT],
            logger,
            outputPath: options?.[OPTIONS.OUTPUT_PATH],
        }
        // @ts-expect-error TODO
        await writeComponent(opts)
        return { code, state: state as State }
    } else {
        return findAssets({
            ...sharedArgs,
            // @ts-expect-error Args need to be pre-validated
            globPattern: options?.[OPTIONS.GLOB_PATTERN],
            nameScheme: options?.[OPTIONS.NAME_SCHEME] as NamingScheme,
            state,
        })
    }
}
