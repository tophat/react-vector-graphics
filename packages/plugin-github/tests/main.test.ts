import { OPTIONS, STATE } from '../src/constants'
import * as findAssets from '../src/findAssets'
import { run } from '../src/main'
import * as writeComponent from '../src/writeComponent'

import { mockComponent, mockOptions, mockSVG, mockState } from './mocks'

import type { State } from '@svgr/core'

describe('main', () => {
    beforeAll(() => {
        jest.spyOn(findAssets, 'default').mockResolvedValue([
            {
                code: mockSVG,
                state: {
                    ...mockState,
                    [STATE.COMPONENT_FILES]: {},
                } as State,
            },
        ])
        jest.spyOn(writeComponent, 'default').mockResolvedValue()
    })
    afterEach(jest.resetAllMocks)
    afterAll(jest.restoreAllMocks)

    it('finds assets if no code passed', async () => {
        const results = run(
            undefined,
            { options: mockOptions, plugins: [] },
            {},
            console,
        )
        await expect(results).resolves.toMatchSnapshot()
    })

    it('writes component files when code passed', async () => {
        const results = run(
            mockComponent,
            { options: mockOptions, plugins: [] },
            mockState,
            console,
        )
        await expect(results).resolves.toMatchSnapshot()
        expect(writeComponent.default).toHaveBeenCalledWith({
            assetFile: mockState[STATE.FILE_PATH as keyof typeof mockState],
            code: mockComponent,
            commitMessagePatterns: {
                create: mockOptions[OPTIONS.COMMIT_CREATE_PATTERN],
                delete: mockOptions[OPTIONS.COMMIT_DELETE_PATTERN],
                update: mockOptions[OPTIONS.COMMIT_UPDATE_PATTERN],
            },
            componentFiles:
                mockState[STATE.COMPONENT_FILES as keyof typeof mockState],
            componentName:
                mockState[STATE.COMPONENT_NAME as keyof typeof mockState],
            componentNameOld:
                mockState[STATE.COMPONENT_NAME_OLD as keyof typeof mockState],
            diffType: mockState[STATE.DIFF_TYPE as keyof typeof mockState],
            fileExt: mockOptions[OPTIONS.FILE_EXT],
            folderPath: mockOptions[OPTIONS.FOLDER_PATH],
            github: {
                api: mockOptions[OPTIONS.API],
                base: mockOptions[OPTIONS.BASE],
                head: mockOptions[OPTIONS.HEAD],
                owner: mockOptions[OPTIONS.OWNER],
                repo: mockOptions[OPTIONS.REPO],
            },
            logger: console,
            outputPath: mockOptions[OPTIONS.OUTPUT_PATH],
        })
    })
})
