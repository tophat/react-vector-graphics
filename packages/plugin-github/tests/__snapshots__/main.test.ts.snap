// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`main finds assets if no code passed 1`] = `
[
  {
    "code": "<svg>mock</svg>",
    "state": {
      "assets/componentFiles": {},
      "componentName": "mockIcon",
      "filePath": "assets/mock.icon.svg",
      "github/diffType": "added",
    },
  },
]
`;

exports[`main writes component files when code passed 1`] = `
{
  "code": "
import React from 'react'
export default function mockIcon() {
    return <svg>mock updated</svg>
}
",
  "state": {
    "assets/componentFiles": {
      "README.md": "## mockIcon

mock usage notes",
    },
    "componentName": "mockIcon",
    "filePath": "assets/mock.icon.svg",
    "github/diffType": "added",
  },
}
`;
